fomantic/Fomantic-UI
# feat(modal): a11y support, tabbing in modal, close icon focus/keypress

## Description
This PR adds accessibility (a11y) support to the modal module

- add proper aria attributes
- make sure tabbing only takes place inside the modal and does not tab to fields hidden by the dimmer
- make a close button a button and tabbable as well

I also separated the closeIcon from the previous close selector which was supposed to be able to also close the modal on each action button when the selector is enhanced.
Reason for separation of closeIcon is the new ability to trigger the closeicon by space or enter which is needed as accessibility requirement and was not possible before.

#### Attention
> The aria attributes are only applied when modal is called dynamically via JS attributes (means without an already prepared modal DOM structure). When using existing Modal DOM nodes, adding proper aria attributes/labels is still the job of the developer as Modal is not adjusting existing DOM node attributes (for example an id has to be created for aria-labelledby).

However, the tabbing feature works independently of existing aria labels and was a general bug anyway.

## Testcase
- Try to tab through the modals fields / buttons
- Try to focus the close icon and press space or enter
- Use a screen reader or check developers tools tab "accessibility" on the generated modal dom node to look for aria labels

### Before
- Tabbing after last action button leads into tabbing something else of your browser or even OS
- Tabbing before the first action button will focus the button behind the dimmer and after that your browser address bar and OS elements
- The close icon cannot be focussed, thus trying to press space or enter is useless and has no effect
- No aria attributes / labels, modal is not recognized in a screen reader properly
- Icon only Button is not recognized by screen reader

https://jsfiddle.net/lubber/7tqo3xkm/15/

### After
- Tabbing only works inside the modal
- Close icon can be focussed
- Pressing space or enter when close icon is focussed will close the modal
- A screen reader will recognize the icon only button as "mail" buttons
- A screen reader (or developer tools, tab "accessibility" will recognize the modal with proper aria attributes/labels

https://jsfiddle.net/lubber/7tqo3xkm/14/

## Screenshot
### Before
![aria_modal_wrong](https://user-images.githubusercontent.com/18379884/128549664-d1920195-a444-4daf-a35b-f66692036344.gif)

### After
![aria_modal_right](https://user-images.githubusercontent.com/18379884/128549674-d6f8d263-1c82-47ba-9690-7d7fbf1794be.gif)

## Closes
https://github.com/Semantic-Org/Semantic-UI/issues/5773
https://github.com/Semantic-Org/Semantic-UI/issues/733

## Related
https://github.com/Semantic-Org/Semantic-UI/issues/348
https://github.com/go-gitea/gitea/issues/7057
